### ISSUE TITLE
Chromium crash when writing a large memfs file

### ISSUE THREAD

**Issue by Beuc** [emscripten-core/emscripten#7630]
Chromium 70 displays a "Aw, Snap!" when writing a file >= 108MB.

```
#include <stdio.h>
#include <stdlib.h>

#define BUFSIZE 1024*1024

int main(void) {
  FILE* f = fopen("/archive.rpa", "w");
  unsigned char buf[BUFSIZE] = "";
  for (int i = 0; i < BUFSIZE; i++) {
    buf[i] = rand() % 256;
  }
  for (int i = 0; i < 150; i++) {  // fails at 108 for me
    fwrite(buf, BUFSIZE, 1, f);
  }
  printf("Wrote file.\n");
}
```

This is consistent whatever the memory options I provide:
```
$ emcc ../testlarge.c -o testlarge.html -s TOTAL_MEMORY=1GB -s ALLOW_MEMORY_GROWTH=1
```

Nothing in the (disconnected) console, but a bit of information on the command line:
```
<--- Last few GCs --->
arking 47 ms) (average mu = 0.915, current mu = 0.978) finalize incremental[70:0x560853e63640]     1612 ms: Mark-sweep 1721.3 (1725.4) -> 1574.0 (1578.1) MB, 5.0 / 0.0 ms  (+ 0.8 ms in 4 steps since start of marking, biggest step 0.6 ms, walltime since start of marking 1216 ms) (average mu = 0.993, current mu = 0.996) allocation[70:0x560853e63640]     1706 ms: Mark-sweep 1574.0 (1578.1) -> 577.5 (581.6) MB, 94.7 / 0.0 ms  (average mu = 0.876, current mu = 0.000) allocation failure GC in old space requested


<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 0x56084ddc86ee]
Security context: 0x06953a021079 <String[21]: http://localhost:8001>
    1: write [0x3225aa7a43a1] [http://localhost:8001/testlarge.js:~2400] [pc=0x30bbd6ae1b21](this=0x3225aa7a1d99 <Object map = 0x273c6e0b3851>,0x07eb43631a91 <JSObject>,0x3225aa7a1d51 <Int8Array map = 0x273c6e0a4f81>,6640,1048576,112197632,0x2ffb7b2025b1 <undefined>)
    2: write [0x3225aa7a1de1] [http://localhost:8001/testlarge.js:4...
```

**Comment by kripken** [comment]
Best to file an issue on the chromium tracker directly, I think: https://bugs.chromium.org/p/chromium/issues/list (and please link to it here if you do).

**Comment by Beuc** [comment]
https://bugs.chromium.org/p/chromium/issues/detail?id=913861

**Comment by juj** [comment]
Chromium has had a longstanding bug where allocating a large ArrayBuffer would not `throw` an exception if it ran out of memory, but instead they would make the whole page AwwSnap - this might be something similar to that. https://bugs.chromium.org/p/chromium/issues/detail?id=42342 , https://bugs.chromium.org/p/chromium/issues/detail?id=257018 , https://bugs.chromium.org/p/chromium/issues/detail?id=536816 are some reports. They have stated it to be fixed various times in the past, and have seen commits go in that do fixes, but I presume the issue is wider and dependent on code flow/context, making the problem appear in multiple different scenarios, which is why the issue keeps reappearing. If you have good deterministic test case build to attach to the bug report, that may be useful.

**Comment by Beuc** [comment]
Thanks for the input.
I did submit the 100%-crash code I posted above, but looking at the bugs you mentioned they seem to already have that kind of test case?

**Comment by towca** [comment]
@Beuc, have you heard anything on the issue, or managed to deal with it somehow?

**Comment by Beuc** [comment]
No :/ You could ping the chromium bug item https://bugs.chromium.org/p/chromium/issues/detail?id=913861 .

**Comment by Beuc** [comment]
@sbc100 upstream's feedback is not optimistic ("probably not something we're likely to be able to support") - can you check if there's a way forward?

**Comment by Beuc** [comment]
@kripken wrt to your answer at chromium, maybe my question is naive, but is there a way to store efficiently (1 byte -> 1 byte) the file in memfs? Either in the memfs internals (like malloc), or at the application level by somehow pre-allocating the file?

(or I could ask users to convert their .zip to a .data+.js - but since I plan to implement compression at the application level due to web hosts limitation, I'd probably get the same issue?)

**Comment by kripken** [comment]
@Beuc well, that's what happens in the typed array case, 1 byte is 1 byte there. We only use normal JS arrays because they can append, so if you do many appends it can be faster. But in v8 such arrays take 8x the space, turns out, so it really is not a good idea for large arrays.

Options here may include:

 * Just removing the `MEMFS_APPEND_TO_TYPED_ARRAYS` and always use typed arrays, even when appending. This may be slow in some cases, but won't fail pathologically in others.
 * Keep the option but switch to it by default, so users can change to the old behavior if desired.

@juj, any objections to removing that option (first of the 2 items here)? I think you may have cared about it.

**Comment by Beuc** [comment]
> This may be slow in some cases

That's what I'd like to avoid, since decompressing the game data on start-up is slow enough already :/
I'll try and time MEMFS_APPEND_TO_TYPED_ARRAYS when I'm done with my other pending issues.

Since RAM is tight (esp. on mobile devices) using it as default sounds good to me.

My question was also: to maintain good perfs, is there a way to use typed array efficiently for generated files (from the C API), or is it something that can only be done from JavaScript (e.g. from the file packager's runtime)? But if things are efficient enough by default maybe we don't care.

**Comment by kripken** [comment]
Not sure why there should be a difference between the C API and JS APIs - maybe I don't know which APIs you mean? Do you mean the C stdio API and the JS FS APIs? (both should end up using the code in library_memfs.js)

**Comment by Beuc** [comment]
To put it differently, is there a way to rewrite this bug's test case so that (efficient) typed arrays are always used?

**Comment by juj** [comment]
> @juj, any objections to removing that option (first of the 2 items here)? I think you may have cared about it.

Sorry, can you rephrase what is being proposed here? Is the intent to make `-s MEMFS_APPEND_TO_TYPED_ARRAYS=1` the default build option, and remove support for `-s MEMFS_APPEND_TO_TYPED_ARRAYS=0` case? That plan does sound ok for me.

However before going down that path, has it been confirmed that using the linker flag `-s MEMFS_APPEND_TO_TYPED_ARRAYS=1` avoids the Chrome crash issue? Scanning the comment history above, I am not sure if that was yet tested?

**Comment by Beuc** [comment]
I checked the test case and my initial program (.zip extraction).
Using MEMFS_APPEND_TO_TYPED_ARRAYS=1 not only stops crashing under Chromium, but is also faster.

Updated testcase with some timing:
```
#include <stdio.h>
#include <stdlib.h>

#define BUFSIZE 1024*1024
#include <sys/time.h>
  

int main(void) {
  FILE* f = fopen("/archive.rpa", "w");
  unsigned char buf[BUFSIZE] = "";
  for (int i = 0; i < BUFSIZE; i++) {
    buf[i] = rand() % 256;
  }

  struct timeval tv1, tv2;
  gettimeofday(&tv1, NULL);

  for (int i = 0; i < 150; i++) {  // fails at 108 for me
    fwrite(buf, BUFSIZE, 1, f);
  }

  gettimeofday(&tv2, NULL);
  printf("Wrote file in %fs.\n",
	 (tv2.tv_sec - tv1.tv_sec) + (tv2.tv_usec - tv1.tv_usec) / 1000000.0);
}
```

Typically 10x faster for the test case, and ~10-40% faster for the .zip extraction.

I suppose writing 8x less in memory has its benefits, even with some reallocation :)

**Comment by kripken** [comment]
Cool, thanks @Beuc!

@juj yes, the plan as you said it, remove the option, and always use typed arrays. I opened #7918 for this.
